service/matching/matching_engine_test.go

```diff
@@ -687,6 +690,119 @@ func (s *matchingEngineSuite) TestPollWorkflowTaskQueues_NamespaceHandover() {
 	s.Equal(common.ErrNamespaceHandover.Error(), err.Error())
 }
 
+func (s *matchingEngineSuite) TestPollActivityTaskQueues_InternalError() {
+	namespaceId := uuid.New()
+	tl := "queue"
+	taskQueue := &taskqueuepb.TaskQueue{Name: "queue", Kind: enumspb.TASK_QUEUE_KIND_NORMAL}
+
+	addRequest := matchingservice.AddActivityTaskRequest{
+		NamespaceId:            namespaceId,
+		Execution:              &commonpb.WorkflowExecution{WorkflowId: "workflowID", RunId: uuid.NewRandom().String()},
+		ScheduledEventId:       int64(5),
+		TaskQueue:              taskQueue,
+		ScheduleToStartTimeout: timestamp.DurationFromSeconds(0),
+	}
+
+	// add an activity task
+	_, _, err := s.matchingEngine.AddActivityTask(context.Background(), &addRequest)
+	s.NoError(err)
+	s.EqualValues(s.taskManager.getTaskCount(newUnversionedRootQueueKey(namespaceId, tl, enumspb.TASK_QUEUE_TYPE_ACTIVITY)), 1)
+
+	// task is dropped with no retry; RecordActivityTaskStarted should only be called once
+	s.mockHistoryClient.EXPECT().RecordActivityTaskStarted(gomock.Any(), gomock.Any(), gomock.Any()).
+		Return(nil, serviceerror.NewInternal("Internal error")).Times(1)
```

Comment:
just curious, does this test actually fail if this returns, say, an Unavailable error? I'm wondering how deeply we need to check the "no retry" part

Reply:
Yeah, the test fails if we return that error. Personally don't think we need to go more deeper, the only place I can think of adding more testing could be the gRPC interceptor but I feel we don't need to given that we are only checking and converting an error type. 

open to hearing your thoughts on this.

Reply:
I think it's fine, it was just a little subtle to think about how the test checks the right thing (e.g. is there a delay before the retry?)